File: apps/api/aqapi/quest/entity/quest_exp_by_level_entity.py
from datetime import datetime
from sqlalchemy import Column, Integer, DateTime, ForeignKey, func, UniqueConstraint, CheckConstraint
from sqlalchemy.orm import relationship, Mapped, mapped_column
from aqapi.core.entity.base_entity import BaseEntity
from aqapi.core.entity.base_translation_entity import BaseTranslationEntity
from aqapi.core.config.db_config import DB_CONF


class QuestExpByLevelEntity(BaseEntity):
    """クエスト経験値(レベル別)エンティティ"""

    __tablename__ = "quest_exp_by_level"
    __table_args__ = (
        # 一意制約
        UniqueConstraint("quest_id", "level", name="uq_quest_exp_by_level_quest_level"),
        # レベルは0より大きい
        CheckConstraint("level > 0", name="chk_quest_exp_by_level_level_positive"),
        # 経験値は0以上
        CheckConstraint("exp >= 0", name="chk_quest_exp_by_level_exp_non_negative"),
    )

    quest_id: Mapped[int] = mapped_column(Integer, ForeignKey("quests.id", ondelete="CASCADE"), nullable=False, comment="クエストID")
    level: Mapped[int] = mapped_column(Integer, nullable=False, comment="レベル")
    exp: Mapped[int] = mapped_column(Integer, nullable=False, comment="必要経験値")

    quest = relationship("QuestsEntity", foreign_keys=[quest_id])

File: packages/backend/src/features/quest/entity/questExpByLevelEntity.ts
import {
  Entity,
  Column,
  ManyToOne,
  JoinColumn,
  Unique,
  Check,
} from "typeorm";
import { BaseTransactionEntity } from "@backend/core/entity/baseTransactionEntity";
import { QuestEntity } from "./questEntity";

/**
 * クエスト経験値(レベル別)エンティティ
 */
@Entity("quest_exp_by_level")
@Unique("uq_quest_exp_by_level_quest_level", ["questId", "level"])
@Check("chk_quest_exp_by_level_level_positive", "level > 0")
@Check("chk_quest_exp_by_level_exp_non_negative", "exp >= 0")
export class QuestExpByLevelEntity extends BaseTransactionEntity {
  @Column({ name: "quest_id", type: "int", nullable: false, comment: "クエストID" })
  questId!: number;
  @Column({ name: "level", type: "int", nullable: false, comment: "レベル" })
  level!: number;
  @Column({ name: "exp", type: "int", nullable: false, comment: "必要経験値" })
  exp!: number;

  @ManyToOne(() => QuestEntity, { nullable: false, onDelete: "CASCADE" })
  @JoinColumn({ name: "quest_id", referencedColumnName: "id", foreignKeyConstraintName: "fk_quest_exp_by_level_quest_id" })
  quest!: QuestEntity;
}
